### PR TITLE
Topic/network simplify

### LINKED ISSUE #1213: virtcontainers: Eliminate legacy networking features
# Eliminate legacy networking models

Prior to the addition of tcMirroring support kata-runtime had compatibility issues with some CNI plugins some of which were addressed by the bridged model. With the addition of tc mode there are no gaps in networking that can be filled by the bridged mode or enlightened mode (which was never implemented).

Eliminate both of these option to simplify the setup.

Also switch the default networking model to tc to ensure broader compatibility.

Also addresses some of the simplification needed for https://github.com/kata-containers/runtime/issues/1113

### PR DESCRIPTION
**Pull request by mcastelino**
# Eliminate legacy networking models

Prior to the addition of tcMirroring support kata-runtime had compatibility issues with some CNI plugins some of which were addressed by the bridged model. With the addition of tc mode there are no gaps in networking that can be filled by the bridged mode or enlightened mode (which was never implemented).

Eliminate both of these option to simplify the setup.

Also addresses some of the simplification needed for #1113 
Fixes #1213

### CODE REVIEW

**Comment by mcastelino** [comment]
@grahamwhaley @devimc @egernst any idea

The CI keeps failing

```
The command "sudo apt update -y -qq" failed and exited with 100 during .
```

**Comment by devimc** [comment]
restarting travis

**Comment by devimc** [comment]
@mcastelino btw

```
# github.com/kata-containers/runtime/virtcontainers
../virtcontainers/network.go:93:35: undefined: NetXConnectTcFilterModel
Makefile:445: recipe for target '/tmp/jenkins/workspace/kata-containers-runtime-ubuntu-16-04-PR-firecracker/go/src/github.com/kata-containers/runtime/kata-runtime' failed
make: *** [/tmp/jenkins/workspace/kata-containers-runtime-ubuntu-16-04-PR-firecracker/go/src/github.com/kata-containers/runtime/kata-runtime] Error 2
Build step 'Execute shell' marked build as failure
```

**Comment by sameo** [comment]
@mcastelino In terms of performance, we know that tc mirroring is superior to the bridged mode?

**Comment by mcastelino** [comment]
> @mcastelino In terms of performance, we know that tc mirroring is superior to the bridged mode?

@sameo tc mirroring is slightly less performant than macvtap. We did not measure tc mirror vs bridged mode. However as macvtap mode should support all the modes that bridged mode supports, even if tcmirror mode is lower performance that bridged mode, the user should not loose when choosing an opinionated configuration.

/cc @amshinde

**Comment by mcastelino** [comment]
/cc @amshinde @sboeuf 

We will also need to update https://github.com/clearcontainers/runtime/blob/master/docs/architecture/architecture.md#networking with the updated networking setup once this PR goes through

**Comment by sameo** [comment]
> > @mcastelino In terms of performance, we know that tc mirroring is superior to the bridged mode?
> 
> @sameo tc mirroring is slightly less performant than macvtap. We did not measure tc mirror vs bridged mode. However as macvtap mode should support all the modes that bridged mode supports, even if tcmirror mode is lower performance that bridged mode, the user should not loose when choosing an opinionated configuration.
>

So having macvtap and tc mirrroring make bridged redundant. Sounds good to me.

**Comment by grahamwhaley** [comment]
Got this on the 16.04 CI - looks like there might be some debug/rework to do here @mcastelino before we re-fire the CIs?
```
=== RUN   TestCreateMacVtap
--- FAIL: TestCreateMacVtap (0.00s)
    <autogenerated>:1: 
                          
	Error Trace:	network_test.go:212
        
	Error:      	Received unexpected error:
        
	            	Unsupported link type bridge
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x10052c6]

goroutine 1072 [running]:
testing.tRunner.func1(0xc000312100)
	/usr/local/go/src/testing/testing.go:792 +0x6a7
panic(0x1168460, 0x1d224e0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/kata-containers/runtime/virtcontainers.TestCreateMacVtap(0xc000312100)
	/tmp/jenkins/workspace/kata-containers-runtime-ubuntu-16-04-PR/go/src/github.com/kata-containers/runtime/virtcontainers/network_test.go:214 +0x206
testing.tRunner(0xc000312100, 0x12a3818)
	/usr/local/go/src/testing/testing.go:827 +0x163
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:878 +0x651
FAIL	github.com/kata-containers/runtime/virtcontainers	4.557s
Makefile:524: recipe for target 'go-test' failed
```

File: virtcontainers/network.go [hunk 1]
```diff
@@ -73,12 +61,8 @@ func (n NetInterworkingModel) IsValid() bool {
 const (
 	defaultNetModelStr = "default"
 
-	bridgedNetModelStr = "bridged"
-
```
Comment (jodh-intel) [inline virtcontainers/network.go]: Won't this require changes to the config files (`internetworking_model` comments)?:

- https://github.com/kata-containers/runtime/blob/master/cli/config/configuration-fc.toml.in
- https://github.com/kata-containers/runtime/blob/master/cli/config/configuration-qemu.toml.in

**Comment by amshinde** [comment]
> We will also need to update https://github.com/clearcontainers/runtime/blob/master/docs/architecture/architecture.md#networking with the updated networking setup once this PR goes through

@mcastelino Yes the docs need to be updated with all the changes we have done recently. Let me know if you are going to take a look at that, I can take a shot at it as well.

**Comment by amshinde** [comment]
> So having macvtap and tc mirrroring make bridged redundant. Sounds good to me.

@sameo @mcastelino Yes, bridged mode has shown the least performance when I measured the performance of each of the modes and all networking drivers are supported between macvtap and tc, making bridged redundant.

File: virtcontainers/network.go (hunk 1 above)
Comment (mcastelino) [reply virtcontainers/network.go]: @jodh-intel I missed that. Let me fix it

**Comment by mcastelino** [comment]
> Got this on the 16.04 CI - looks like there might be some debug/rework to do here @mcastelino before we re-fire the CIs?

@grahamwhaley I love our CI. Let me retest this with macvtap any my end.

**Comment by mcastelino** [comment]
@amshinde looks like our unit test case exercises path that our code will no longer take. i.e. we create the macvtap on top of bridge that serves as the parent link. I eliminated the use of bridge interface here.

File: virtcontainers/network.go (hunk 1 above)
Comment (mcastelino) [reply virtcontainers/network.go]: fixed

File: virtcontainers/qemu_arch_base.go [hunk 2]
```diff
@@ -432,15 +432,8 @@ func (q *qemuArchBase) appendVSockPCI(devices []govmmQemu.Device, vsock kataVSOC
 
 func networkModelToQemuType(model NetInterworkingModel) govmmQemu.NetDeviceType {
 	switch model {
-	case NetXConnectBridgedModel:
-		return govmmQemu.MACVTAP //TODO: We should rename MACVTAP to .NET_FD
 	case NetXConnectMacVtapModel:
 		return govmmQemu.MACVTAP
-	//case ModelEnlightened:
```
Comment (mcastelino) [inline virtcontainers/qemu_arch_base.go]: Note: This is already implemented implicitly today. The interconnect model is only used today when the interface cannot be directly connected to the VM. The vhost-user interface is also handed using a special case and works with only a certain class of vhost-user plugins.

So our default model is actually ModelEnlightened when running in tcFilter or macvtap mode. The none mode is the only one that does not behave this way.

**Comment by devimc** [comment]
@mcastelino CI is not happy

```
 rpc error: code = Unknown desc = failed to start sandbox container: failed to create containerd task: OCI runtime create failed: Unknown type bridged: unknown
```

**Comment by caoruidong** [comment]
+1 for this. We can simply our network code. Many different types of links, endpoints and network models.

**Comment by jodh-intel** [comment]
@mcastelino - any update on this?

**Comment by grahamwhaley** [comment]
/test
let's see if the CIs are any happier - and if not, let's grab some log fragments to help @mcastelino figure out why not....

**Comment by jodh-intel** [comment]
Ping @kata-containers/runtime - input required over here!

**Comment by amshinde** [comment]
Logs dont seem to be available any more. Retriggering..
/test

**Comment by jodh-intel** [comment]
Half the CI's are still failing on this PR @mcastelino.

**Comment by amshinde** [comment]
@mcastelino Code still seems to access bridged mode somewhere. Can you take a look.
Here is what the CI failures look like
```
Failure in Spec Setup (BeforeEach) [5.198 seconds]
[k8s.io] Container
/tmp/jenkins/workspace/kata-containers-runtime-ubuntu-18-04-PR/go/src/github.com/kubernetes-sigs/cri-tools/pkg/framework/framework.go:72
  runtime should support basic operations on container
  /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-18-04-PR/go/src/github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:68
    runtime should support stopping container [Conformance] [BeforeEach]
    /tmp/jenkins/workspace/kata-containers-runtime-ubuntu-18-04-PR/go/src/github.com/kubernetes-sigs/cri-tools/pkg/validate/container.go:100

    failed to create PodSandbox: rpc error: code = Unknown desc = failed to start sandbox container: failed to create containerd task: Unknown type bridged: unknown
    Expected error:
        <*status.statusError | 0xc0004868d0>: {
            Code: 2,
            Message: "failed to start sandbox container: failed to create containerd task: Unknown type bridged: unknown",
            Details: nil,
        }
        rpc error: code = Unknown desc = failed to start sandbox container: failed to create containerd task: Unknown type bridged: unknown
    not to have occurred
```

**Comment by mcastelino** [comment]
@amshinde looking at that failure now.

**Comment by WeiZhang555** [comment]
One question:

Are we sure there's no user using bridge model in some productions currently? If we remove this in next release, will this make them angry?

The code is easy to understand and it looks good to me.

**Comment by jodh-intel** [comment]
Related: https://github.com/kata-containers/runtime/issues/1113.

**Comment by raravena80** [comment]
@mcastelino @amshinde ping, any updates? Thx!

**Comment by raravena80** [comment]
@mcastelino @amshinde nudge

**Comment by grahamwhaley** [comment]
@mcastelino @amshinde 
- did you ever get updates for this to fix the CI fails
- the CI logs are long gone - please re-trigger if/when you get back to this PR and need fresh logs
- what did we think about @WeiZhang555 query of 'do we know we have no bridge customers at present'? I guess one option would be to make this an 'experimental' (off by default) feature for one release, and then disabled in the release after that. In fact, is that not our generally agreed Kata workflow for deprecating features across two releases... @egernst

**Comment by jodh-intel** [comment]
@mcastelino - any update on this?

/cc @amshinde.

**Comment by raravena80** [comment]
@mcastelino @amshinde any updates? Thx.

**Comment by egernst** [comment]
Stepping back - I'd like to verify bridge is not being utilized, first.  We *may* want to set this to be a 2.0 feature.

**Comment by egernst** [comment]
Adding label "needs-decision" -- let's put this on mailing list, and add to next weeks architecture committe call.

Happy to see this removed, but want to get input on whether this will break our users (making it a 2.0 milestone target).

**Comment by raravena80** [comment]
@mcastelino any decision progress? Should we add this to the next arch meeting?

**Comment by jodh-intel** [comment]
@amshinde - could you possibly take over this PR maybe?

**Comment by amshinde** [comment]
@jodh-intel Will do.

**Comment by amshinde** [comment]
@jodh-intel PR updated.

**Comment by amshinde** [comment]
@egernst @jodh-intel PTAL
Lets try to merge this, before this goes stale again.

**Comment by grahamwhaley** [comment]
Fedora CI fail looked like a net/repo issue (failed to update) - re-triggered...

**Comment by amshinde** [comment]
Thanks @grahamwhaley. All tests have passed now.

**Comment by grahamwhaley** [comment]
@amshinde - need one more ack - given it's heavy network nature, and the original author is mcastelino, I'll be happy if you ack this PR, and then we can merge :-)

**Comment by amshinde** [comment]
ping @mcastelino @egernst

**Comment by grahamwhaley** [comment]
re-ping @mcastelino @egernst - CI is happy here, just needs knowledgeable reviewers...

**Comment by amshinde** [comment]
Adding "do-not-merge" until we decide to remove "bridged". This feature has been marked as deprecated in a separate PR.

**Comment by amshinde** [comment]
@kata-containers/runtime @egernst This PR is ready for review now, since we plan to remove `bridged` mode in 1.10. PTAL.

**Comment by amshinde** [comment]
cc @devimc @jodh-intel

**Comment by jschintag** [comment]
I don't see a problem for s390x.

**Pull request merged**